### DISCUSSION TITLE
Support customize type checker in observable.

### DISCUSSION THREAD

**Discussion by janryWang** [mobxjs/mobx#2809] [Ideas]
For some special types(such as ReactNode,Moment,Immutable Object etc.) in deep observable, I hope the typer checker can be customized.
here is the source code location
https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/api/observable.ts#L125
Usage can be like this.
```tsx

makeObservable({
  data:{...}
},{
  data:observable.deep((target,targetPropName)=>{
    if(isReactElement(target) || isMoment(target)) return observable.ref(target)
    return observable(target)
  })
})

```

**Comment by janryWang** [comment]
@mweststrate @FredyC Have any thoughts?

**Comment by mweststrate** [comment]
I never needed it, as in all my use cases observable.ref sufficed. Not
entirely unsympathetic to the idea, but wondering if it shouldn't be more
of a global config then, doing it per declaration might be error prone as
it is easy to forget, and also requires mobx to carry context around about
the decorator that was used. Curious about real life use cases indeed if
.ref / .shallow didn't suffice?

**Reply by janryWang** [reply]
Thanks reply @mweststrate
For example, I have an node tree state for dynamic render UI. I need mobx to manage the whole tree state in deep.
```ts
interface ObservableTreeNode {
    tagName:string
    props:{
       [key:string]?:string | object | any[] | Moment | ReactNode | Immutable | OtherTypes //I don’t know which property is a react node or a normal object
    }
    children?:ObservableTreeNode[]
}
interface MyObservableState{
  tree:ObservableTreeNode
}
```

But observe can not recognized ReactNode It will automatically recursively traverse the react node and pollute the structure of the react node. This is something I don't expect to happen, so I hope to ignore the type of react node.

**Reply by urugator** [reply]
Imo you should have some action, like `setNodeValue(node, value)` or `addChild(node, value)` or similar that checks the value type and converts it to observable as needed - if the value is plain object you may want to traverse the object and convert it to tree nodes.

**Reply by janryWang** [reply]
I actually just want to dynamically ignore some special data types, when I use observable.deep

**Reply by urugator** [reply]
I now what you want, but we don't support it and I don't think we should. As I said all state mutations are under your controll. You can create whatever abstraction around your observable structure and convert things as needed inside actions. If `deep` doesn't fit your use case, you just shouldn't use it and should build the tree manually, based on your conversion rules.

**Reply by janryWang** [reply]
According to your theory, shouldn't observe.deep exist? Because it is also a black box, it is actually difficult to perceive what data structure is inside. Now I just want this black box to handle some details more accurately.

**Reply by urugator** [reply]
`observable.deep` is a conviniency, something that fits most of the cases and assumes plain data structures. It's not intended to work with any arbitrary data structure. Imo generally mobx should aim to work in two "modes": 
explicit - user is in full control and responsbile for making things observable, batched etc.
implicit - provides some convinient defaults, that are guaranteed to work in our domain - meaning everything is observable, batched etc.
I don't think it's worth to make this automatic implicit behavior further configurable, because you can always achieve the same with the explicit behavior and it often significantly increases complexity of the library - more than the complexity you would have to introduce to you code.
I also worry that this is one of the features that would lead to missuse and antipatterns. It already seems to me that some ppl treat mobx as some kind of AOP library, which can inject arbitrary logic into any object (not to mention they should often use simple composition instead).
Lastly we would expose something that was so far implementation detail (maintenance burden) and we would probably have to adapt API so there aren't multiple ways of doing same thing.

**Reply by janryWang** [reply]
Thank you for your patient response, but if custom type checker is not supported, it is difficult for me to build an observable tree model. If there is a some hacked solution, I would choose to use a hack, but the official does not want to expose such an API. ,and  I can’t use hacked solution, it’s really hard for me now.

**Reply by janryWang** [reply]
Two options I can think of:
1. Support custom type checker, this is the most direct
2. Support custom annotations, I can directly create an annotation similar to observable.deep

**Reply by mweststrate** [reply]
as @urugator indicated you can write your own to-observable logic that does recursively exactly the same, something along the lines of:

```javascript
function createMyFancyObservable(value) {
  if (!value || typeof value !== "object") return value; // primitive
  if (isReactNode(value) || momentJSThing(value) || /* etc */) return value; // no conversion
  if (Array.isArray(value)) return observable.array(value.map(createMyFancyObservable), { deep: false })
  if (value instanceof Map) return observable.map(Array.from(value.entries()).map(([key, value] => [key, createMyFancyObservable(value)]), { deep: false }))
  return observable.object(Object.fromEntries(Object.entries(value).map((key, value) => [key, createMyFancyObservable(value)]), { deep: false })
}
```

**Reply by janryWang** [reply]
```tsx
function createMyFancyObservable(value) {
  if (!value || typeof value !== "object") return value; // primitive
  if (isValidElement(value)) return value; // no conversion
  if (Array.isArray(value))
    return observable.array(value.map(createMyFancyObservable), {
      deep: false
    });
  const target = Object.fromEntries(
    Object.entries(value).map((key, value) => [
      key,
      createMyFancyObservable(value)
    ])
  );
  return observable.object(target, { deep: false });
}

const aa = createMyFancyObservable({});

aa.bb = {
  cc: {
    dd: <div>Hello world</div>
  }
};
```

Thanks Reply, But this example still doesn't seem to recognize ReactNode. @mweststrate

This is [codesandbox](https://codesandbox.io/s/hungry-bird-c52sb?file=/src/App.js)

In the console, I will see that the react element will also be converted into a proxy object

**Comment by iChenLei** [comment]
I think this idea only work on in initialization stage.
```ts
const data = createMyFancyObservable(rawData);

// not
const data = createMyFancyObservable({});
data = rawData;
```
Rerun the script
```js
function createMyFancyObservable(value) {
  if (!value || typeof value !== "object") return value; // primitive
  if (isValidElement(value)) return value; // no conversion
  if (Array.isArray(value))
    return observable.array(value.map(createMyFancyObservable), {
      deep: false
    });
  const target = Object.fromEntries(
    Object.entries(value).map([key, value] => [
      key,
      createMyFancyObservable(value)
    ])
  );
  return observable.object(target, { deep: false });    👈🏻   // <- still covert object property to observable 
 }

createMyFancyObservable({
  bb: {
    cc: {
      dd: <div>Hello world</div>
    }
  }
})
```
![2021-02-18 15 18 10](https://user-images.githubusercontent.com/14012511/108319548-857a1580-71fc-11eb-8387-e5103a2be8c0.png)

but not work, cause `{ depp: false }` not work as expected?

![2021-02-18 15 15 22](https://user-images.githubusercontent.com/14012511/108319289-23211500-71fc-11eb-813a-0e93aa931fcb.png)

### append
from `mweststrate`'s comment
```js
const data = createMyFancyObservable({});
data = createMyFancyObservable(rawData);  // reassign
```

**Reply by janryWang** [reply]
I hope it not only supports initial creation, but also supports dynamic creation of observable objects. 😅

**Comment by mweststrate** [comment]
Pass new values through the same func before you assign them to the tree

**Reply by janryWang** [reply]
So I have to do a deep recursion on all the data every time, the performance will become very bad.

**Comment by mweststrate** [comment]
The recursion is exactly the same regardless whether MobX does it or you do
it yourself. But you only need to apply it to new values you are about to
store. Not to everything every time.
